src/Execution/Execution.php

```diff
@@ -52,10 +47,57 @@ public function execute(
             return new ExecutionResult(null, [$error]);
         }
 
-        $data   = $this->createExecutor($context, $errorHandler)->execute();
-        $errors = $context->getErrors();
+        $valuesResolver = new ValuesResolver();
+        $fieldCollector = new FieldCollector($context, $valuesResolver);
+
+        $data = $this->executeOperation($operationName, $context, $fieldCollector, $valuesResolver);
+
+        if ($data instanceof PromiseInterface) {
+            $data->then(function ($resolvedData) use (&$data) {
+                $data = $resolvedData;
+            });
+        }
+
+        return new ExecutionResult($data, $context->getErrors());
+    }
+
+    /**
+     * @param null|string      $operationName
+     * @param ExecutionContext $context
+     * @param FieldCollector   $fieldCollector
+     * @param ValuesResolver   $valuesResolver
+     * @return array|mixed|null|PromiseInterface
+     */
+    protected function executeOperation(
+        ?string $operationName,
+        ExecutionContext $context,
+        FieldCollector $fieldCollector,
+        ValuesResolver $valuesResolver
+    ) {
+        $strategy = $operationName === 'mutation'
+            ? new SerialExecutionStrategy($context, $fieldCollector, $valuesResolver)
+            : new ParallelExecutionStrategy($context, $fieldCollector, $valuesResolver);
+
+        $result = null;
+
+        try {
+            $result = $strategy->execute();
+        } catch (ExecutionException $exception) {
+            $context->addError($exception);
+        } catch (\Throwable $exception) {
+            $context->addError(
+                new ExecutionException($exception->getMessage(), null, null, null, null, null, $exception)
+            );
+        }
+
+        if ($result instanceof PromiseInterface) {
```

Comment:
In https://sergeyzhuk.me/2018/01/16/reactphp-managing-promises/ they mention "I don’t know exactly what the resolver will give me", i.e. it seems it's easy to wrap any result in a promise. This could potentially simplify the execution logic a bit, don't you think?

Reply:
It could, but I don't how that would affect the performance though.

Reply:
I doubt it affects performance

Reply:
@Jalle19 Do you want to give it a try?

Reply:
What do you mean?

Reply:
@Jalle19 Changing the code to always use promises.